# **Pandas - Analyzing DataFrames**

In [ ]:
import pandas as pd

**Viewing the Data**
* One of the most used method for getting a quick overview of the DataFrame, is the head() method.
* The head() method returns the headers and a specified number of rows, starting from the top.

In [ ]:
# Get the DataFrame:
df = pd.read_csv('/content/Iris.csv')
print(df.head(10))

In [ ]:
# Get a quick overview by printing the first 10 rows of the DataFrame:
df = pd.read_csv('/content/Iris.csv')
print(df.head(10))

   Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa
5   6            5.4           3.9            1.7           0.4  Iris-setosa
6   7            4.6           3.4            1.4           0.3  Iris-setosa
7   8            5.0           3.4            1.5           0.2  Iris-setosa
8   9            4.4           2.9            1.4           0.2  Iris-setosa
9  10            4.9           3.1            1.5           0.1  Iris-setosa


In [ ]:
# Print the first 5 rows of the DataFrame:
df = pd.read_csv('/content/Iris.csv')
print(df.head())

   Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa


In [ ]:
# Print the last 5 rows of the DataFrame:
print(df.tail()) 

      Id  SepalLengthCm  ...  PetalWidthCm         Species
145  146            6.7  ...           2.3  Iris-virginica
146  147            6.3  ...           1.9  Iris-virginica
147  148            6.5  ...           2.0  Iris-virginica
148  149            6.2  ...           2.3  Iris-virginica
149  150            5.9  ...           1.8  Iris-virginica

[5 rows x 6 columns]


**Info About the Data**
* The DataFrames object has a method called info(), that gives you more information about the data set.

In [ ]:
# Print information about the data:
print(df.info()) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             150 non-null    int64  
 1   SepalLengthCm  150 non-null    float64
 2   SepalWidthCm   150 non-null    float64
 3   PetalLengthCm  150 non-null    float64
 4   PetalWidthCm   150 non-null    float64
 5   Species        150 non-null    object 
dtypes: float64(4), int64(1), object(1)
memory usage: 7.2+ KB
None


**Example-2**

In [ ]:
df = pd.read_csv('/data.csv')
print(df.head())

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0


In [ ]:
print(df.info()) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB
None


**Null Values**
* The info() method also tells us how many Non-Null values there are present in each column, and in our data set it seems like there are 164 of 169 Non-Null values in the "Calories" column.
* Which means that there are 5 rows with no value at all, in the "Calories" column, for whatever reason.
* Empty values, or Null values, can be bad when analyzing data, and you should consider removing rows with empty values. This is a step towards what is called cleaning data, and you will learn more about that in the next tutorials.

In [ ]:
# get dataframe
df = pd.read_csv('/content/Iris.csv')
print(df)

      Id  SepalLengthCm  ...  PetalWidthCm         Species
0      1            5.1  ...           0.2     Iris-setosa
1      2            4.9  ...           0.2     Iris-setosa
2      3            4.7  ...           0.2     Iris-setosa
3      4            4.6  ...           0.2     Iris-setosa
4      5            5.0  ...           0.2     Iris-setosa
..   ...            ...  ...           ...             ...
145  146            6.7  ...           2.3  Iris-virginica
146  147            6.3  ...           1.9  Iris-virginica
147  148            6.5  ...           2.0  Iris-virginica
148  149            6.2  ...           2.3  Iris-virginica
149  150            5.9  ...           1.8  Iris-virginica

[150 rows x 6 columns]


In [ ]:
# Find Maximum value from each colums
print(df.max())

Id                          150
SepalLengthCm               7.9
SepalWidthCm                4.4
PetalLengthCm               6.9
PetalWidthCm                2.5
Species          Iris-virginica
dtype: object


In [ ]:
# Find Maximum value from second colums
df['SepalLengthCm'].max()

7.9

In [ ]:
# Find Maximum value from second colums
df[['SepalLengthCm','PetalWidthCm']].max()

SepalLengthCm    7.9
PetalWidthCm     2.5
dtype: float64

In [ ]:
# Find the Mean value
print(df.mean())

Id               75.500000
SepalLengthCm     5.843333
SepalWidthCm      3.054000
PetalLengthCm     3.758667
PetalWidthCm      1.198667
dtype: float64


In [ ]:
# Median
print(df.median())

Id               75.50
SepalLengthCm     5.80
SepalWidthCm      3.00
PetalLengthCm     4.35
PetalWidthCm      1.30
dtype: float64


In [ ]:
# Mode
print(df.mode())

      Id  SepalLengthCm  ...  PetalWidthCm          Species
0      1            5.0  ...           0.2      Iris-setosa
1      2            NaN  ...           NaN  Iris-versicolor
2      3            NaN  ...           NaN   Iris-virginica
3      4            NaN  ...           NaN              NaN
4      5            NaN  ...           NaN              NaN
..   ...            ...  ...           ...              ...
145  146            NaN  ...           NaN              NaN
146  147            NaN  ...           NaN              NaN
147  148            NaN  ...           NaN              NaN
148  149            NaN  ...           NaN              NaN
149  150            NaN  ...           NaN              NaN

[150 rows x 6 columns]


In [ ]:
# Sum of each Column
print(df.sum())

Id                                                           11325
SepalLengthCm                                                876.5
SepalWidthCm                                                 458.1
PetalLengthCm                                                563.8
PetalWidthCm                                                 179.8
Species          Iris-setosaIris-setosaIris-setosaIris-setosaIr...
dtype: object


In [ ]:
# Square and Sum of each column
print(df.pow(2).sum())

TypeError: ignored

In [ ]:
df1 = pd.read_csv('/data.csv')
print(df.head())

   Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa


In [ ]:
# Square and Sum of each column
print(df1.pow(2).sum())

Duration      989500.00
Pulse        1986981.00
Maxpulse     3082172.00
Calories    34725999.02
dtype: float64


In [ ]:
# Cube and Sum of each column
print(df1.pow(3).sum())

Duration    1.375445e+08
Pulse       2.218457e+08
Maxpulse    4.258639e+08
Calories    3.118391e+10
dtype: float64


##**Pandas - Data Correlations**


**Finding RelationshipsFinding Relationships**
* A great aspect of the Pandas module is the corr() method.
* The corr() method calculates the relationship between each column in your data set.
* The examples in this page uses a CSV file called: 'data.csv'.

In [ ]:
df.corr()

Note: The corr() method ignores "not numeric" columns.

**Result Explained**
* The Result of the corr() method is a table with a lot of numbers that represents how well the relationship is between two columns.
* The number varies from -1 to 1.
* 1 means that there is a 1 to 1 relationship (a perfect correlation), and for this data set, each time a value went up in the first column, the other one went up as well.
* 0.9 is also a good relationship, and if you increase one value, the other will probably increase as well.
* -0.9 would be just as good relationship as 0.9, but if you increase one value, the other will probably go down.
* 0.2 means NOT a good relationship, meaning that if one value goes up does not mean that the other will.

**Perfect Correlation:**
* We can see that "Duration" and "Duration" got the number 1.000000, which makes sense, each column always has a perfect relationship with itself.

**Good Correlation:**
* "Duration" and "Calories" got a 0.922721 correlation, which is a very good correlation, and we can predict that the longer you work out, the more calories you burn, and the other way around: if you burned a lot of calories, you probably had a long work out.

**Bad Correlation:**
* "Duration" and "Maxpulse" got a 0.009403 correlation, which is a very bad correlation, meaning that we can not predict the max pulse by just looking at the duration of the work out, and vice versa.

## **Data Filteration in Pandas**

**Python iloc() function**

* Python iloc() function enables us to select a particular cell of the dataset, that is, it helps us select a value that belongs to a particular row or column from a set of values of a data frame or dataset.
* With iloc() function, we can retrieve a particular value belonging to a row and column using the index values assigned to it.
* Remember, iloc() function accepts only integer type values as the index values for the values to be accessed and displayed.

**Syntax**
* iloc**(** start_row **:** end_row **,** start_col **:** end_col **)**

In [ ]:
df = pd.read_csv('/data.csv')
print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

In [ ]:
# Display Record Number 3 to 15 for all column i.e in row index 2 to row index 14 
df = pd.read_csv('/data.csv')
print(df.iloc[2:14])

    Duration  Pulse  Maxpulse  Calories
2         60    103       135     340.0
3         45    109       175     282.4
4         45    117       148     406.0
5         60    102       127     300.0
6         60    110       136     374.0
7         45    104       134     253.3
8         30    109       133     195.1
9         60     98       124     269.0
10        60    103       147     329.3
11        60    100       120     250.7
12        60    106       128     345.3
13        60    104       132     379.3


In [ ]:
# Display all record of Duration and pulse
df = pd.read_csv('/data.csv')
print(df.iloc[:,0:2])

     Duration  Pulse
0          60    110
1          60    117
2          60    103
3          45    109
4          45    117
..        ...    ...
164        60    105
165        60    110
166        60    115
167        75    120
168        75    125

[169 rows x 2 columns]


In [ ]:
# Display all record of column Maxpulse and calories
df = pd.read_csv('/data.csv')
print(df.iloc[:,2:4])

     Maxpulse  Calories
0         130     409.1
1         145     479.0
2         135     340.0
3         175     282.4
4         148     406.0
..        ...       ...
164       140     290.8
165       145     300.0
166       145     310.2
167       150     320.4
168       150     330.4

[169 rows x 2 columns]


In [ ]:
# Display first 20 records of first Three Column
df = pd.read_csv('/data.csv')
print(df.iloc[0:20,0:3])

    Duration  Pulse  Maxpulse
0         60    110       130
1         60    117       145
2         60    103       135
3         45    109       175
4         45    117       148
5         60    102       127
6         60    110       136
7         45    104       134
8         30    109       133
9         60     98       124
10        60    103       147
11        60    100       120
12        60    106       128
13        60    104       132
14        60     98       123
15        60     98       120
16        60    100       120
17        45     90       112
18        60    103       123
19        45     97       125


In [ ]:
# Print all record of Column "Duration" and "Maxpulse" Column i.e Column index 0 and Column index 2
df = pd.read_csv('/data.csv')
print(df.iloc[:,[0, 3]])

     Duration  Calories
0          60     409.1
1          60     479.0
2          60     340.0
3          45     282.4
4          45     406.0
..        ...       ...
164        60     290.8
165        60     300.0
166        60     310.2
167        75     320.4
168        75     330.4

[169 rows x 2 columns]


In [ ]:
# Print the record index of 2,7,8,29,30 of column index 0 and Clomun index 3
df = pd.read_csv('/data.csv')
print(df.iloc[[2,7,8,29,30],[0, 3]])

    Duration  Calories
2         60     340.0
7         45     253.3
8         30     195.1
29        60     380.3
30        60     243.0


In [ ]:
# select rows with particular indexes and particular columns
df.iloc[[0,2],[1,3]]

Pulse  Calories
0    110     409.1
2    103     340.0

**loc in Pandas**
* loc is label-based, which means that we have to specify the name of the rows and columns that we need to filter out.
* For example, let’s say we search for the rows whose index is 1, 2 or 100. We will not get the first, second or the hundredth row here. Instead, we will get the results only if the name of any index is 1, 2 or 100.
* So, we can filter the data using the loc function in Pandas even if the indices are not an integer in our dataset.

 

**iloc in Pandas**
* On the other hand, iloc is integer index-based. So here, we have to specify rows and columns by their integer index.
* Let’s say we search for the rows with index 1, 2 or 100. It will return the first, second and hundredth row, regardless of the name or labels we have in the index in our dataset.
* We will see multiple examples in this article on how to use loc and iloc for the data selection and data update process in Python.

 

In [ ]:
# display the record of Row index 29
df = pd.read_csv('/data.csv')
print(df.loc[29])

Duration     60.0
Pulse       102.0
Maxpulse    129.0
Calories    380.3
Name: 29, dtype: float64


In [ ]:
data = {
        "Protein": [38, 27, 18, 25, 29],
        "Fats": [32, 39, 40, 35, 33],
        "Carbs": [46, 47, 50, 60, 65]
       }

df1 = pd.DataFrame(data, index = ["P1", "P2", "P3", "P4", "P5"])

print(df1) 

    Protein  Fats  Carbs
P1       38    32     46
P2       27    39     47
P3       18    40     50
P4       25    35     60
P5       29    33     65


In [ ]:
# print the record where row index is P3 to P5
print(df1.loc['P3':'P5'])

    Protein  Fats  Carbs
P3       18    40     50
P4       25    35     60
P5       29    33     65


In [ ]:
# select all rows with a condition
df1.loc[df1.Fats >= 35]

Protein  Fats  Carbs
P2       27    39     47
P3       18    40     50
P4       25    35     60

In [ ]:
# select with multiple conditions
df1.loc[(df1.Protein >= 12) & (df1.Carbs >= 50)]

Protein  Fats  Carbs
P3       18    40     50
P4       25    35     60
P5       29    33     65

In [ ]:
#slice
df.loc[1:3]

Duration  Pulse  Maxpulse  Calories
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4

In [ ]:
# select few columns with a condition
df1.loc[(df1.Fats >= 35), ['Fats', 'Carbs']]

Fats  Carbs
P2    39     47
P3    40     50
P4    35     60

In [ ]:
# update a column with condition
df1.loc[(df1.Fats >= 12), ['Protein']] = 50
df1

Protein  Fats  Carbs
P1       50    32     46
P2       50    39     47
P3       50    40     50
P4       50    35     60
P5       50    33     65

In [ ]:
# update multiple columns with condition
df1.loc[(df1.Fats >= 40), ['Protein', 'Carbs']] = ['55','58']
df1

Protein  Fats Carbs
P1      50    32    46
P2      50    39    47
P3      55    40    58
P4      50    35    60
P5      50    33    65